[Validator] Use PHP attributes when creating custom validation constraints

yoannrenard opened the pull request:
Describe how to use PHP attributes when creating custom validation constraints

yoannrenard committed:
[Validator] Use PHP attributes when creating custom validation constraints

OskarStark commented on validation/custom_constraint.rst:
```diff
@@ -12,20 +12,37 @@ alphanumeric characters.
+
+        use Symfony\Component\Validator\Constraint;
+
+        #[\Attribute()]
```
@derrabus do we recommend a leading slash and the brackets?

yoannrenard replied:
@OskarStark 
I got rid of the brackets like it is in the php annotations snippets.  
Though the leading slash looks required to me as it doesn't work without.

derrabus replied:
@OskarStark Symfony's CS policy is that we don't import classes from the global namespace, so the leading backslash is correct.

OskarStark commented on validation/custom_constraint.rst:
```diff
@@ -128,6 +145,25 @@ You can use custom validators like the ones provided by Symfony itself:
+        {
+            // ...
+
+            #[Assert\NotBlank()]
+            #[AcmeAssert\ContainsAlphanumeric()]
```
```suggestion
            #[Assert\NotBlank]
            #[AcmeAssert\ContainsAlphanumeric]
```

OskarStark commented on validation/custom_constraint.rst:
```diff
@@ -241,6 +277,14 @@ not to the property:
 
+    .. code-block:: php-attributes
+
+        #[AcmeAssert\ProtocolClass()]
```
```suggestion
        #[AcmeAssert\ProtocolClass]
```

derrabus commented:
The PR should target 5.2 and we should add a note that attribute constraints have been added with Symfony 5.2.

yoannrenard commented:
I've added a small note inspired by the one made for [creating routes as attributes](https://symfony.com/doc/current/routing.html#creating-routes-as-attributes-or-annotations) as you requested @derrabus and changed the PR target to 5.2

derrabus reviewed (approved):
Looks good. Shall we add a section explaining how to enable a constraint for both usages (Doctrine Annotations + PHP Attributes)? Or is this a problem, an application usually does not have?

OskarStark commented on validation/custom_constraint.rst:
```diff
@@ -12,20 +12,43 @@ alphanumeric characters.
 
-    // src/Validator/ContainsAlphanumeric.php
-    namespace App\Validator;
+.. versionadded:: 5.2
```
This should be moved below the configuration block, thanks

yoannrenard replied:
Done @OskarStark

yoannrenard commented:
@derrabus You mean something explaining how to use both (Doctrine annotations + PHP attributes) at the same time? Like using a constraint that supports attributes and one which doesn't?

derrabus commented:
You can build classes that are a valid Doctrine annotation as well as a valid PHP attribute. Almost all of Symfony's built-in constraints work that way. The examples you provided only work with one of the annotation implementations. But that's probably knowledge most developers won't need.

I think, we're good here.

xabbuh commented on validation/custom_constraint.rst:
```diff
@@ -12,20 +12,43 @@ alphanumeric characters.
+
+.. versionadded:: 5.2
+
+    The ability to use PHP attributes to configure constraint was introduced in
```
[...] to configure constraint**s** [...]

yoannrenard replied:
fixed @xabbuh

OskarStark commented:
Thank you Yoann.
